### PR TITLE
feat: add verify decorator to fastify

### LINKED ISSUE #63: Allow Anonymous Access
## 🚀 Feature Proposal

Allow anonymous access

## Motivation

Make the authorization header optional via config
I took a look at fastify-auth and plugin scopes but cannot see a way to achive this

## Example

```js
'use strict'

const fastify = require('fastify')()
const bearerAuthPlugin = require('fastify-bearer-auth')
const keys = new Set(['a-super-secret-key', 'another-super-secret-key'])

fastify.register(bearerAuthPlugin, {keys, allowAnonymous: true})

fastify.get('/foo', (req, reply) => {
  reply.send("Public Endpoint")
})

fastify.listen({port: 8000}, (err) => {
  if (err) {
    fastify.log.error(err.message)
    process.exit(1)
  }
  fastify.log.info('http://127.0.0.1:8000/foo')
})
```

Example Changes 
```js
if (!header) {
     if (allowAnonymous) {
          return next();
     }
      const noHeaderError = Error('missing authorization header')
      fastifyReq.log.error('unauthorized: %s', noHeaderError.message)
      if (contentType) fastifyRes.header('content-type', contentType)
      fastifyRes.code(401).send(errorResponse(noHeaderError))
      return
    }
```

### PR DESCRIPTION
**Pull request by climba03003**
Fixes: #63 

Changes:
- add options `legacy` to add `onRequest` hooks as before
- add `verifyBearerAuth` to `fastify` to allow simple use-case
- add `verifyBearerAuthFactory` to `fastify` to allow complex use-case
- add example for simple integration

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Comment by climba03003** [comment]
Note: This is a `semver-major` change as it do not add the `onRequest` hook as before.

File: test/integration.test.js [hunk 1]
```diff
@@ -52,3 +52,74 @@ test('missing header route fails correctly', (t) => {
...
+      fastify.verifyBearerAuth
+    ]),
+    handler: function (_, reply) {
+      reply.send({ hello: 'world' })
+    }
+  })
+
+  await fastify.ready()
+
+  test('anonymous should pass', async (t) => {
```
Comment (jsumners) [inline test/integration.test.js:81]: If this is supposed to be a sub-test, then:
```suggestion
  t.test('anonymous should pass', async (t) => {
```

File: test/integration.test.js [hunk 2]
```diff
@@ -52,3 +52,74 @@ test('missing header route fails correctly', (t) => {
...
+    try {
+      const res = await fastify.inject({ method: 'GET', url: '/anonymous' })
+      t.strictEqual(res.statusCode, 200)
+      t.match(JSON.parse(res.body).hello, 'world')
+    } catch (err) {
+      t.error(err)
+    }
+  })
+
+  test('bearer auth should pass', async (t) => {
```
Comment (jsumners) [inline test/integration.test.js:92]:
```suggestion
  t.test('bearer auth should pass', async (t) => {
```

File: test/integration.test.js [hunk 3]
```diff
@@ -52,3 +52,74 @@ test('missing header route fails correctly', (t) => {
...
+        }
+      })
+      t.strictEqual(res.statusCode, 200)
+      t.match(JSON.parse(res.body).hello, 'world')
+    } catch (err) {
+      t.error(err)
+    }
+  })
+
+  test('bearer auth should fail', async (t) => {
```
Comment (jsumners) [inline test/integration.test.js:109]:
```suggestion
  t.test('bearer auth should fail', async (t) => {
```

**Review by jsumners — CHANGES_REQUESTED** [review]
I am against this change. The point of this plugin is to provide " a simple request hook for the Fastify web framework" as stated in the opening of the readme. The idea is that one adds the plugin to the encapsulation scope they need to protect and it does what is necessary at the right hook point (which should really be `preValidation` since that hook is now available). This simple plugin should not require the usage of another plugin, `fastify-auth`, in order to be usable.

File: Readme.md [hunk 4]
```diff
@@ -56,6 +56,7 @@ sent to the client (optional)
    rejects, or throws, a HTTP status of `500` will be sent. `req` is the Fastify
    request object. If `auth` is a function, `keys` will be ignored. If `auth` is
    not a function, or `undefined`, `keys` will be used.
+* `legacy`: turn on legacy mode to add bearer auth for `onRequest` hook.
```
Comment (mcollina) [inline Readme.md:59]: let's call this `addHook` instead.

**Review by mcollina — CHANGES_REQUESTED** [review]
Good work!

**Comment by climba03003** [comment]
> 
> 
> I am against this change. The point of this plugin is to provide " a simple request hook for the Fastify web framework" as stated in the opening of the readme. The idea is that one adds the plugin to the encapsulation scope they need to protect and it does what is necessary at the right hook point (which should really be `preValidation` since that hook is now available). This simple plugin should not require the usage of another plugin, `fastify-auth`, in order to be usable.

I change the behavior back to `addHook` as default and you have your ability to disable this behavior as you want. You may also use the provided `decorator` for the complex use case.

File: Readme.md [hunk 5]
```diff
@@ -81,6 +83,30 @@ a `{error: message}` body; no further request processing will be performed.
 [fplugin]: https://github.com/fastify/fastify/blob/master/docs/Plugins.md
 [prehook]: https://github.com/fastify/fastify/blob/master/docs/Hooks.md
 
+## Integration
```
Comment (jsumners) [inline Readme.md:86]:
```suggestion
## Integration with `fastify-auth`
```

File: Readme.md [hunk 6]
```diff
@@ -81,6 +83,30 @@ a `{error: message}` body; no further request processing will be performed.
...
+
+This plugin can integrate with `fastify-auth`, you can follow the example below.
```
Comment (jsumners) [inline Readme.md:88]:
```suggestion
This plugin can integrate with `fastify-auth` by following this example:
```

File: Readme.md [hunk 7]
```diff
@@ -81,6 +83,30 @@ a `{error: message}` body; no further request processing will be performed.
...
+
+```JS
```
Comment (jsumners) [inline Readme.md:90]: Consistency nit:
```suggestion
```js
```

File: plugin.js [hunk 8]
```diff
@@ -90,13 +89,21 @@ function compare (a, b) {
...
 }
 
-function plugin (fastify, options, next) {
-  fastify.addHook('onRequest', factory(options))
-  next()
+function plugin (fastify, options, done) {
+  options = Object.assign({ addHook: true }, options)
+
+  fastify.decorate('verifyBearerAuthFactory', verifyBearerAuthFactory)
+  fastify.decorate('verifyBearerAuth', verifyBearerAuthFactory(options))
```
Comment (jsumners) [inline plugin.js:96]: If `addHook` is `true`, should we still add these decorations?

File: plugin.js [hunk 9]
```diff
@@ -90,13 +89,21 @@ function compare (a, b) {
...
+
+  if (options.addHook) {
```
Comment (jsumners) [inline plugin.js:98]:
```suggestion
  if (options.addHook === true) {
```

**Review by jsumners — COMMENTED** [review]
I am okay with the latest revision of this PR. It retains the original functionality as the default functionality while providing compatibility with `fastify-auth`. A few minor corrections and a question left to be answered.

File: plugin.js (hunk 8 above)
Comment (climba03003) [reply plugin.js:96]: You are right, if the `hook` is added. There are no way people needing those decorations as they cannot override the verification.

File: plugin.js [hunk 10]
```diff
@@ -90,13 +89,21 @@ function compare (a, b) {
...
-function plugin (fastify, options, next) {
-  fastify.addHook('onRequest', factory(options))
-  next()
+function plugin (fastify, options, done) {
+  options = Object.assign({ addHook: true }, options)
+
+  if (options.addHook === true) {
+    fastify.addHook('onRequest', verifyBearerAuthFactory(options))
+  } else {
+    fastify.decorate('verifyBearerAuthFactory', verifyBearerAuthFactory)
```
Comment (cliedeman) [inline plugin.js:98]: what is the value in exposing this as a decorator instead of exporting the function directly?

File: plugin.js (hunk 8 above)
Comment (cliedeman) [reply plugin.js]: Always adding the decorators is better for the typescript definitions

**Review by mcollina — APPROVED** [review]
lgtm

File: plugin.js (hunk 10 above)
Comment (climba03003) [reply plugin.js:98]: You can access it through `fastify.verifyBearerAuthFactory` and `fastify.verifyBearerAuth`. or even `this.verifyBearerAuthFactory` and `this.verifyBearerAuth`.

Comment (climba03003) [reply plugin.js:98]: FYI, you can access the `factory` function without this PR.
```JS
import { internals } from 'fastify-bearer-auth`
const verifyBearerAuthFactory = internals.factory
```

**Review by jsumners — APPROVED** [review]
LGTM

**Pull request merged**